design/20200331.acme-self-check-failure-policy.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Failure policy for ACME self checks
+authors:
+  - "@anton-johansson"
+reviewers: []
+approvers: []
+creation-date: 2020-03-31
+last-updated: 2020-03-31
+status: ?
+---
+
+# Failure policy for ACME self checks
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [API Changes](#api-changes)
+  - [Controller Behaviour](#controller-behaviour)
+  - [Internal API Resource Behaviour](#internal-api-resource-behaviour)
+  - [Test Plan](#test-plan)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+      - [Alpha](#alpha)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+      - [Removing a deprecated flag](#removing-a-deprecated-flag)
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+When requesting certificates from Lets Encrypt, cert-manager performs a self-check to make sure that we have properly configured the ACME challenge response for the domain. This is to avoid redundant calls to Lets Encrypt and avoid getting rate limited. This is very good safety net. However, it would be a great addition to cert-manager if this check could be disabled or at least configured to allow failure.
+
+## Motivation
+
+In certain cluster setups, the ACME challenge self-check performed by cert-manager isn't always possible. For example, when using MetalLB and you have local DNS records that points towards the IP address assigned by MetalLB you might run into some hairpin issues, causing the traffic to fail.
+
+There is an issue describing some of the problems: [#1292](https://github.com/jetstack/cert-manager/issues/1292)
+
+### Goals
+
+* Add support for allowing the self-check to fail
+
+## Proposal
+
+We need a new policy enum to control the failure policy. It should be specified on the ACME solver level, so it works for both DNS01 and HTTP01. The two values in the initial version would be:
+* `Fail` - the self check must pass before the request is sent.
+* `Ignore` - the request will be sent even if the self check fails.
+
+### API changes
+
+To specify how ACME challenge self-checks are allowed to fail, we need to slightly modify the API.
+
+```go
+package v1
+
+// Describes how the ACME challenge self-check behaves when it fails.
+// +kubebuilder:validation:Enum=Fail;Ignore
+type ACMEFailurePolicy string
+
+const (
+	// The default failure policy. This policy causes the request to fail before it is sent to the certificate provider.
+	ACMEFailurePolicyFail ACMEFailurePolicy = "Fail"
```

Comment:
Hm, come to think, we don't actually implement `Fail` behaviour right now, we effectively have a `WaitForever` policy...

Reply:
Yeah, that is true. The `Certificate` resource will be stuck on `Pending`, right?

Reply:
Changed to `WaitForever` and improved some documentation.